File: ios/PretextTokenization.swift
import Foundation

internal func tokenize(
    _ text: String,
    style: NativeTextStyle
) -> [NativeTokenDescriptor] {
    var tokens: [NativeTokenDescriptor] = []
    var current = ""
    var currentStart = -1
    var currentEnd = -1
    var mode: TokenMode?
    let nsText = text as NSString
    var cursor = 0

    func flushCurrent() {
        guard !current.isEmpty, currentStart >= 0, currentEnd >= 0 else {
            return
        }

        tokens.append(
            NativeTokenDescriptor(
                text: current,
                startUTF16: currentStart,
                endUTF16: currentEnd,
                style: style
            )
        )
        current = ""
        currentStart = -1
        currentEnd = -1
    }

    while cursor < nsText.length {
        let range = nsText.rangeOfComposedCharacterSequence(at: cursor)
        let segment = nsText.substring(with: range)

        if segment == newlineToken {
            flushCurrent()
            mode = nil
            tokens.append(
                NativeTokenDescriptor(
                    text: newlineToken,
                    startUTF16: range.location,
                    endUTF16: range.location + range.length,
                    style: style
                )
            )
            cursor = range.location + range.length
            continue
        }

        let nextMode: TokenMode = segment.unicodeScalars.allSatisfy {
            CharacterSet.whitespacesAndNewlines.contains($0) && !isUnicodeNewline($0)
        } ? .whitespace : .text

        if mode == nextMode {
            current.append(segment)
            currentEnd = range.location + range.length
        } else {
            flushCurrent()
            mode = nextMode
            current = segment
            currentStart = range.location
            currentEnd = range.location + range.length
        }

        cursor = range.location + range.length
    }

    flushCurrent()
    return tokens
}

internal func tokenizeBreakUnits(
    _ text: String,
    style: NativeTextStyle
) -> [NativeTokenDescriptor] {
    let nsText = text as NSString
    guard nsText.length > 0 else {
        return []
    }

    var tokens: [NativeTokenDescriptor] = []
    var cursor = 0
    while cursor < nsText.length {
        let range = nsText.rangeOfComposedCharacterSequence(at: cursor)
        tokens.append(
            NativeTokenDescriptor(
                text: nsText.substring(with: range),
                startUTF16: range.location,
                endUTF16: range.location + range.length,
                style: style
            )
        )
        cursor = range.location + range.length
    }
    return tokens
}

internal func appendInlineSegment(
    _ segment: InlineSegment,
    resolvedStyle: NativeTextStyle,
    tokens: inout [NativeTokenDescriptor],
    breakUnits: inout [NativeTokenDescriptor]
) {
    let baseOffset = min(tokens.last?.endUTF16 ?? Int.max, breakUnits.last?.endUTF16 ?? Int.max)
    let resolvedBaseOffset = baseOffset == Int.max ? 0 : baseOffset

    if segment.breakBehavior.lowercased() == breakBehaviorNever {
        appendNeverBreakTokens(
            segment.text,
            baseOffset: resolvedBaseOffset,
            style: resolvedStyle,
            output: &tokens
        )
        appendNeverBreakTokens(
            segment.text,
            baseOffset: resolvedBaseOffset,
            style: resolvedStyle,
            output: &breakUnits
        )
        return
    }

    tokens += tokenize(segment.text, style: resolvedStyle).map { token in
        NativeTokenDescriptor(
            text: token.text,
            startUTF16: token.startUTF16 + resolvedBaseOffset,
            endUTF16: token.endUTF16 + resolvedBaseOffset,
            style: token.style
        )
    }
    breakUnits += tokenizeBreakUnits(segment.text, style: resolvedStyle).map { token in
        NativeTokenDescriptor(
            text: token.text,
            startUTF16: token.startUTF16 + resolvedBaseOffset,
            endUTF16: token.endUTF16 + resolvedBaseOffset,
            style: token.style
        )
    }
}

private func appendNeverBreakTokens(
    _ text: String,
    baseOffset: Int,
    style: NativeTextStyle,
    output: inout [NativeTokenDescriptor]
) {
    let nsText = text as NSString
    var localStart = 0

    while localStart < nsText.length {
        let remainingRange = NSRange(location: localStart, length: nsText.length - localStart)
        let nextNewline = nsText.range(of: newlineToken, options: [], range: remainingRange)

        if nextNewline.location == NSNotFound {
            output.append(
                NativeTokenDescriptor(
                    text: nsText.substring(with: remainingRange),
                    startUTF16: baseOffset + localStart,
                    endUTF16: baseOffset + nsText.length,
                    style: style
                )
            )
            break
        }

        if nextNewline.location > localStart {
            output.append(
                NativeTokenDescriptor(
                    text: nsText.substring(with: NSRange(location: localStart, length: nextNewline.location - localStart)),
                    startUTF16: baseOffset + localStart,
                    endUTF16: baseOffset + nextNewline.location,
                    style: style
                )
            )
        }

        output.append(
            NativeTokenDescriptor(
                text: newlineToken,
                startUTF16: baseOffset + nextNewline.location,
                endUTF16: baseOffset + nextNewline.location + nextNewline.length,
                style: style
            )
        )
        localStart = nextNewline.location + nextNewline.length
    }
}

private func isUnicodeNewline(_ scalar: Unicode.Scalar) -> Bool {
    CharacterSet.newlines.contains(scalar)
}

File: android/src/main/java/com/margelo/nitro/pretext/PretextTokenization.kt
package com.margelo.nitro.pretext

import java.text.BreakIterator

internal fun tokenize(text: String, style: NativeTextStyle): List<NativeTokenDescriptor> {
  val tokens = ArrayList<NativeTokenDescriptor>()
  val current = StringBuilder()
  var currentStart = -1
  var currentEnd = -1
  var mode: TokenMode? = null
  var cursor = 0

  fun flushCurrent() {
    if (current.isEmpty() || currentStart < 0 || currentEnd < 0) {
      return
    }

    tokens += NativeTokenDescriptor(
      text = current.toString(),
      start = currentStart,
      end = currentEnd,
      style = style,
    )
    current.setLength(0)
    currentStart = -1
    currentEnd = -1
  }

  while (cursor < text.length) {
    val codePoint = text.codePointAt(cursor)
    val charCount = Character.charCount(codePoint)
    val nextCursor = cursor + charCount
    val segment = text.substring(cursor, nextCursor)

    if (codePoint == '\n'.code) {
      flushCurrent()
      mode = null
      tokens += NativeTokenDescriptor(
        text = NEWLINE_TOKEN,
        start = cursor,
        end = nextCursor,
        style = style,
      )
      cursor = nextCursor
      continue
    }

    val nextMode = if (Character.isWhitespace(codePoint)) TokenMode.WHITESPACE else TokenMode.TEXT
    if (mode == nextMode) {
      current.append(segment)
      currentEnd = nextCursor
    } else {
      flushCurrent()
      mode = nextMode
      currentStart = cursor
      currentEnd = nextCursor
      current.append(segment)
    }

    cursor = nextCursor
  }

  flushCurrent()
  return tokens
}

internal fun tokenizeBreakUnits(text: String, style: NativeTextStyle): List<NativeTokenDescriptor> {
  if (text.isEmpty()) {
    return emptyList()
  }

  return collectGraphemeBoundariesForText(text, style.locale)
    .zipWithNext()
    .mapNotNull { (start, end) ->
      if (end <= start) {
        null
      } else {
        NativeTokenDescriptor(
          text = text.substring(start, end),
          start = start,
          end = end,
          style = style,
        )
      }
    }
}

internal fun appendInlineSegmentTokens(
  segment: InlineSegment,
  resolvedStyle: NativeTextStyle,
  baseOffset: Int,
  tokens: MutableList<NativeTokenDescriptor>,
  breakUnits: MutableList<NativeTokenDescriptor>,
) {
  val breakBehavior = segment.breakBehavior.lowercase()
  if (breakBehavior == BREAK_BEHAVIOR_NEVER) {
    appendNeverBreakTokens(segment.text, baseOffset, resolvedStyle, tokens)
    appendNeverBreakTokens(segment.text, baseOffset, resolvedStyle, breakUnits)
    return
  }

  tokens += tokenize(segment.text, resolvedStyle).map { token ->
    NativeTokenDescriptor(
      text = token.text,
      start = token.start + baseOffset,
      end = token.end + baseOffset,
      style = token.style,
    )
  }
  breakUnits += tokenizeBreakUnits(segment.text, resolvedStyle).map { token ->
    NativeTokenDescriptor(
      text = token.text,
      start = token.start + baseOffset,
      end = token.end + baseOffset,
      style = token.style,
    )
  }
}

internal fun collectGraphemeBoundariesForText(
  text: String,
  locale: String = "",
): List<Int> {
  if (text.isEmpty()) {
    return listOf(0)
  }

  val breaker = BreakIterator.getCharacterInstance(resolveTextLocale(locale))
  breaker.setText(text)
  val rawBoundaries = ArrayList<Int>()
  var boundary = breaker.first()
  while (boundary != BreakIterator.DONE) {
    rawBoundaries += boundary
    boundary = breaker.next()
  }
  return rawBoundaries
    .filter { candidate -> candidate == 0 || candidate == text.length || !isUnsafeGraphemeBoundary(text, candidate) }
    .distinct()
    .sorted()
    .let { boundaries ->
      when {
        boundaries.firstOrNull() != 0 -> listOf(0) + boundaries
        boundaries.lastOrNull() != text.length -> boundaries + text.length
        else -> boundaries
      }
    }
}

private fun appendNeverBreakTokens(
  text: String,
  baseOffset: Int,
  style: NativeTextStyle,
  output: MutableList<NativeTokenDescriptor>,
) {
  var localStart = 0
  while (localStart <= text.length) {
    val newlineIndex = text.indexOf('\n', localStart)
    if (newlineIndex == -1) {
      if (localStart < text.length) {
        output += NativeTokenDescriptor(
          text = text.substring(localStart),
          start = baseOffset + localStart,
          end = baseOffset + text.length,
          style = style,
        )
      }
      break
    }

    if (newlineIndex > localStart) {
      output += NativeTokenDescriptor(
        text = text.substring(localStart, newlineIndex),
        start = baseOffset + localStart,
        end = baseOffset + newlineIndex,
        style = style,
      )
    }
    output += NativeTokenDescriptor(
      text = NEWLINE_TOKEN,
      start = baseOffset + newlineIndex,
      end = baseOffset + newlineIndex + 1,
      style = style,
    )
    localStart = newlineIndex + 1
    if (localStart == text.length) {
      break
    }
  }
}

private fun isUnsafeGraphemeBoundary(text: String, boundary: Int): Boolean {
  if (boundary <= 0 || boundary >= text.length) {
    return false
  }

  if (Character.isHighSurrogate(text[boundary - 1]) || Character.isLowSurrogate(text[boundary])) {
    return true
  }

  val before = text.codePointBefore(boundary)
  val after = text.codePointAt(boundary)
  return before == ZERO_WIDTH_JOINER ||
    after == ZERO_WIDTH_JOINER ||
    isVariationSelector(after) ||
    isEmojiModifier(after) ||
    isUnsafeRegionalIndicatorBoundary(text, boundary) ||
    isIndicVirama(before) ||
    isIndicVirama(after) ||
    isCombiningMark(after)
}

private fun isUnsafeRegionalIndicatorBoundary(text: String, boundary: Int): Boolean {
  if (!isRegionalIndicator(text.codePointBefore(boundary)) || !isRegionalIndicator(text.codePointAt(boundary))) {
    return false
  }

  var regionalIndicatorCount = 0
  var cursor = boundary
  while (cursor > 0) {
    val codePoint = text.codePointBefore(cursor)
    if (!isRegionalIndicator(codePoint)) {
      break
    }
    regionalIndicatorCount += 1
    cursor -= Character.charCount(codePoint)
  }
  return regionalIndicatorCount % 2 == 1
}
